File: library/src/main/java/com/iamkurtgoz/easystore/EasyStoreBuilder.java
package com.iamkurtgoz.easystore;

import android.content.Context;
import android.content.ContextWrapper;
import android.content.SharedPreferences;

public class EasyStoreBuilder {

    private static EasyStoreBuilder easyStoreBuilder;
    private String preferenceName;
    private Context context;
    private int preferenceMode = ContextWrapper.MODE_PRIVATE;
    public static SharedPreferences sharedPreferences;

    public static EasyStoreBuilder from(Context context, String preferenceName, int preferenceMode){
        return new EasyStoreBuilder(context, preferenceName, preferenceMode);
    }

    private EasyStoreBuilder(Context context, String preferenceName, int preferenceMode) {
        this.context = context;
        this.preferenceName = preferenceName;
        this.preferenceMode = preferenceMode;
    }

    public void create() {
        if (context == null) {
            throw new RuntimeException("Context is a null");
        }

        if (preferenceMode == -1) {
            preferenceMode = ContextWrapper.MODE_PRIVATE;
        }

        sharedPreferences = getContext().getSharedPreferences(getPreferenceName(), getPreferenceMode());
        easyStoreBuilder = this;
    }

    public static EasyStoreBuilder getBuilder(){
        return easyStoreBuilder;
    }

    public Context getContext() {
        return context;
    }

    public int getPreferenceMode() {
        return preferenceMode;
    }

    public String getPreferenceName() {
        return preferenceName;
    }

    public SharedPreferences getSharedPreferences() {
        return sharedPreferences;
    }
}

File: library/src/main/java/com/iamkurtgoz/easystore/EasyStoreBuilder.kt
package com.iamkurtgoz.easystore

import android.content.Context

class EasyStoreBuilder {

    companion object {

        @Volatile private var instance: EasyStore? = null
        private val lock = Any()

        operator fun invoke(context: Context, preferenceName: String = "default_preference_name", easyStoreMode: EasyStoreMode = EasyStoreMode.MODE_PRIVATE) : EasyStore = instance ?: synchronized(lock) {
            instance ?: createEasyStore(context, preferenceName, easyStoreMode).also {
                instance = it
            }
        }

        private fun createEasyStore(context: Context, preferenceName: String?, easyStoreMode: EasyStoreMode) : EasyStore {
            return EasyStore.getInstance().init(context, preferenceName, easyStoreMode)
        }
    }

}